apecs/src/Apecs/Stores.hs

```diff
@@ -66,6 +70,71 @@ instance MonadIO m => ExplMembers m (Map c) where
   {-# INLINE explMembers #-}
   explMembers (Map ref) = liftIO$ U.fromList . M.keys <$> readIORef ref
 
+-- | A collection for unit-like types («tags») based on 'Data.IntSet'.
+--
+-- The only types that can be stored are those that share representation with `()`:
+-- i.e. single constructor without parameters: @data Bullet = Bullet@.
+--
+-- A bit more compact than "Map", @O(log(n))@ for most operations, but "reading" is free.
+--
+-- @
+-- cmap $ \(Bullet, Position pos, Velocity vel) -> Position (pos + vel)
+-- @
+--
+-- @Bullet@ will be used for its members query, pre-filtering the rest of the components,
+-- but unlike "Map", the actual value would not use a lookup.
+newtype Tags c = Tags (IORef IntSet)
+
+type instance Elem (Tags c) = c
+
+instance (MonadIO m) => ExplInit m (Tags c) where
+  explInit =
+    liftIO . fmap Tags $
+      newIORef mempty
+
+instance (MonadIO m) => ExplGet m (Tags c) where
+  {-# INLINE explExists #-}
+  explExists (Tags ref) entityId =
+    liftIO .
+      fmap (IntSet.member entityId) $
+        readIORef ref
+
+  {-# INLINE explGet #-}
+  explGet _tags _entityId =
+    {-
+      XXX: Given that explGet is partial in other stores
+      the system should not call explGet without a checking first,
+      or obtaining an entity id from explMembers.
+      So, no need to re-check again and it is okay to produce
+      the tag ex nihilo.
+    -}
+    pure $ unsafeCoerce ()
```

Comment:
This can be combined with compile-time proof using Generics:

```haskell
instance (MonadIO m, G.Generic c, G.Rep c ~ G.D1 m1 (G.C1 m2 G.U1)) => ExplGet m (Tags c) where
```
But will require users to import, enable and derive Generic for their tags.

Reply:
Oops. The m1 and m2 are left ambiguous and can't be covered with forall {m1} {m2}.